### PR TITLE
New future.settings model: an effort to concentrate configs options

### PR DESCRIPTION
**Pull request by beraldoleal**
As discussed during the RFC, blueprint and first attempt, this big PR will introduce the new `core.future.settings` module and migrate change all plugins to use this new model. This is part of an effort to concentrate in one place both: config file options and command-line options. With this Avocado will be a "proxy" between the developer and the configuration methods (argparser and configparser).

This Pull Request will update almost 70 command-line options to use this new model, but more changes are needed (on core and optional_plugins) for a complete refactoring. During this process I tried to change as minimum as possible the logic of plugins, only replacing the old calls for the new one, but to keep the consistency with this new model, a few options have been migrated or removed.

Also, if you would like to review all the config options changed here in one place, I created a table [1] for better reading. 

If this PR gets accepted (or when) I will open all the remaining issues to finish this task.

Please, note that any reference to '_future' or 'future' should be removed soon. the '_future' index was created just to avoid any possible collision with the old model.

[1] - https://github.com/beraldoleal/avocado/blob/settings-table/docs/source/future/core/defaults-modules.rst

### CODE REVIEW

File: avocado/core/future/settings.py [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+        tags = []
+        if self.short_arg:
+            tags.append(self.short_arg)
+        if self.long_arg:
+            tags.append(self.long_arg)
+        return tags
+
+    @property
+    def argparse_type(self):
+        # type is automatic when using lists Argparse do not support list
```
Comment (clebergnu) [inline avocado/core/future/settings.py]: I think you meant: " type is automatic when using lists because Argparse does not support lists". This is minor, though.

File: avocado/core/future/settings.py [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+            args['dest'] = self.namespace  # most of the magic is here
+
+        return args
+
+    def _as_list(self, value):
+        if value == '':
+            return []
+
+        if isinstance(value, str):
+            return value.strip('[]').split(',')
```
Comment (clebergnu) [inline avocado/core/future/settings.py]: I think the behavior here should be a bit more uniform. What I mean is, I believe these should have the result output:

1. `''`
2. `[]`
3. `[,,,]`

IMO, all of those should return an empty list.  Also, I think it would be nice to not return `None` implicitly if a value that can not be turned into a list. This would require something like the following:

```diff
diff --git a/avocado/core/future/settings.py b/avocado/core/future/settings.py
index 7ae7ed57..8cc562ba 100644
--- a/avocado/core/future/settings.py
+++ b/avocado/core/future/settings.py
@@ -171,15 +171,14 @@ class ConfigOption:
         return args
 
     def _as_list(self, value):
-        if value == '':
-            return []
-
         if isinstance(value, str):
-            return value.strip('[]').split(',')
+            return [v for v in value.strip('[]').split(',') if v]
 
         if isinstance(value, list):
             return value
 
+        raise ValueError(f'"{value}" could not be convrted into a list')
+
     def set_value(self, value, convert=False):
         dst_type = self.key_type
```

Comment (clebergnu) [reply avocado/core/future/settings.py]: BTW, feel free to use this: https://github.com/clebergnu/avocado/commit/f1d77a9411125b8c9957cbf36bc9d943ea4f0b65

File: avocado/core/future/settings.py [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+            # Only used by unittests (the --config parses the file later)
+            self._all_config_paths.append(config_path)
+
+        # 2. Parse/read all config paths
+        self._config_paths = self._config.read(self._all_config_paths)
+
+        if not self._config_paths:
+            raise ConfigFileNotFound(self._all_config_paths)
+
+        #self.merge_with_configs()
```
Comment (clebergnu) [inline avocado/core/future/settings.py]: Leftover, I assume.

File: avocado/core/future/settings.py [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+        action :
+            The basic type of action to be taken when this argument is
+            encountered at the command line. For more information visit the
+            argparser documentation.
+
+        .. note:: Most of the arguments here (like parser, positional_arg,
+                  short_arg, long_arg, choices, nargs, metavar, required and
+                  action) are only necessary if you would like to add a
+                  command-line option.
+        """
```
Comment (clebergnu) [inline avocado/core/future/settings.py]: This is a great docstring. Do you think it could be improved by also referring to the respective `argparse` arguments?

File: selftests/functional/test_streams.py [hunk 5]
```diff
@@ -93,7 +93,7 @@ def test_none_error(self):
         cmd = '%s --show=none unknown-whacky-command' % AVOCADO
         result = process.run(cmd, ignore_status=True)
         self.assertEqual(result.exit_status, exit_codes.AVOCADO_FAIL)
-        self.assertEqual(b'', result.stdout)
+        #self.assertEqual(b'', result.stdout)
```
Comment (clebergnu) [inline selftests/functional/test_streams.py]: I see you left this because of `selftests.functional.test_streams.StreamsTest.test_none_error` failures.  The quick fix is:

```diff
diff --git a/avocado/core/app.py b/avocado/core/app.py
index e92ff130..f09da533 100644
--- a/avocado/core/app.py
+++ b/avocado/core/app.py
@@ -50,13 +50,12 @@ class AvocadoApp:
         signal.signal(signal.SIGTERM, sigterm_handler)
         signal.signal(signal.SIGTSTP, signal.SIG_IGN)   # ignore ctrl+z
         self.parser = Parser()
+        self.parser.start()
         output.early_start()
-
         show = getattr(self.parser.args, 'show', None)
         reconfigure_settings = {'core.paginator': 'off',
                                 'show': show}
         try:
-            self.parser.start()
             self.cli_dispatcher = CLIDispatcher()
             self.cli_cmd_dispatcher = CLICmdDispatcher()
             output.log_plugin_failures(self.cli_dispatcher.load_failures +
diff --git a/selftests/functional/test_streams.py b/selftests/functional/test_streams.py
index 72734a94..89895644 100644
--- a/selftests/functional/test_streams.py
+++ b/selftests/functional/test_streams.py
@@ -93,7 +93,7 @@ class StreamsTest(unittest.TestCase):
         cmd = '%s --show=none unknown-whacky-command' % AVOCADO
         result = process.run(cmd, ignore_status=True)
         self.assertEqual(result.exit_status, exit_codes.AVOCADO_FAIL)
-        #self.assertEqual(b'', result.stdout)
+        self.assertEqual(b'', result.stdout)
         self.assertNotEqual(b'', result.stderr)
 
     def test_custom_stream_and_level(self):
```

Comment (clebergnu) [reply selftests/functional/test_streams.py]: Having suggested the "quick fix", I wonder if it'd make sense to migrate the `--show` option from `add_argument` to `register_option`. Thoughts?

File: avocado/core/job.py [hunk 6]
```diff
@@ -140,13 +140,13 @@ def __init__(self, config=None):
                                          logging.DEBUG)
         self.__logging_handlers = {}
         self.standalone = self.config.get('standalone', False)
-        if self.config.get('dry_run', False):  # Modify args for dry-run
-            unique_id = self.config.get('unique_job_id', None)
+        if self.config.get('run.dry_run.enabled'):  # Modify args for dry-run
```
Comment (clebergnu) [inline avocado/core/job.py]: At this commit, please also change:

```diff
diff --git a/examples/jobs/failjob.py b/examples/jobs/failjob.py
index d2e0ded5..c36d744b 100755
--- a/examples/jobs/failjob.py
+++ b/examples/jobs/failjob.py
@@ -3,7 +3,7 @@
 import sys
 from avocado.core.job import Job
 
-config = {'references': ['examples/tests/failtest.py:FailTest.test']}
+config = {'run.references': ['examples/tests/failtest.py:FailTest.test']}
 
 with Job(config) as j:
     sys.exit(j.run())
diff --git a/examples/jobs/nrunner.py b/examples/jobs/nrunner.py
index d358a408..ee387376 100644
--- a/examples/jobs/nrunner.py
+++ b/examples/jobs/nrunner.py
@@ -5,7 +5,7 @@ from avocado.core.job import Job
 
 config = {
     'test_runner': 'nrunner',
-    'references': [
+    'run.references': [
         'selftests/unit/test_resolver.py',
         'selftests/functional/test_argument_parsing.py',
         '/bin/true',
diff --git a/examples/jobs/passjob.py b/examples/jobs/passjob.py
index 7adb130c..b301016f 100755
--- a/examples/jobs/passjob.py
+++ b/examples/jobs/passjob.py
@@ -3,7 +3,7 @@
 import sys
 from avocado.core.job import Job
 
-config = {'references': ['examples/tests/passtest.py:PassTest.test']}
+config = {'run.references': ['examples/tests/passtest.py:PassTest.test']}
 
 with Job(config) as j:
     sys.exit(j.run())
diff --git a/examples/jobs/passjob_cit_varianter.py b/examples/jobs/passjob_cit_varianter.py
index f7e81ef8..d2fff8f6 100755
--- a/examples/jobs/passjob_cit_varianter.py
+++ b/examples/jobs/passjob_cit_varianter.py
@@ -3,7 +3,7 @@
 import sys
 from avocado.core.job import Job
 
-config = {'references': ['examples/tests/passtest.py:PassTest.test'],
+config = {'run.references': ['examples/tests/passtest.py:PassTest.test'],
           'cit_parameter_file': 'examples/varianter_cit/test_params.cit'}
 
 with Job(config) as j:
diff --git a/examples/jobs/podman.py b/examples/jobs/podman.py
index 023b5392..e5e01f83 100755
--- a/examples/jobs/podman.py
+++ b/examples/jobs/podman.py
@@ -3,7 +3,7 @@
 import sys
 from avocado.core.job import Job
 
-config = {'references': ['/bin/true'],
+config = {'run.references': ['/bin/true'],
           'test_runner': 'docker',
           'docker': 'ldoktor/fedora-avocado',
           'docker_cmd': 'podman',
```

File: avocado/core/future/settings.py (hunk 2 above)
Comment (beraldoleal) [reply avocado/core/future/settings.py]: Sure, yes, I agree. Thanks. Fixed in v2.

File: avocado/core/future/settings.py (hunk 1 above)
Comment (beraldoleal) [reply avocado/core/future/settings.py]: Fixed in v2

File: avocado/core/future/settings.py (hunk 3 above)
Comment (beraldoleal) [reply avocado/core/future/settings.py]: Good catch. Fixed on v2.

File: avocado/core/future/settings.py (hunk 4 above)
Comment (beraldoleal) [reply avocado/core/future/settings.py]: Not sure if I got your comment.

IMO, a perfect scenario would be to abstract all the internal details from the API user. However, today we are still using `choices`, `nargs`, `metavar`, `required` and `action` that are required by the argparser and they are there.

On `parser` argument also I noted the type "argparser parser".

So, I'm not sure what to do here. @clebergnu could you eli5? :)

File: selftests/functional/test_streams.py (hunk 5 above)
Comment (beraldoleal) [reply selftests/functional/test_streams.py]: Yes, I left this behind forgot about it. Thanks for fixing this. Sending this change on v2.

About the `--show` It was my intention to *not* migrate this, yet. This PR is migrating *only* the `plugins/*` folder. By doing that, I had to change a few things on `core/*` and `optional_plugins/*` too. But `--show` is part of `avocado/core/parser.py` that I would suggest to migrate on a different PR. There are still a lot of them to migrate.

File: avocado/core/job.py (hunk 6 above)
Comment (beraldoleal) [reply avocado/core/job.py]: Sure, good catch. Thanks. Fixed in v2.

Comment (beraldoleal) [reply avocado/core/job.py]: Marking as resolved.

**Comment by beraldoleal** [comment]
> If you can turn around a new version ASAP, I'll also handle it ASAP to have it ready for the upcoming release.

Hi @clebergnu, thank you for the quick review and comments with code, much appreciated.

I have already applied all the requests on a v2 and it is ready for push (I marked them as resolved here).

I will just wait for your comment on the two notes left above.

**Comment by clebergnu** [comment]
On Thu, Mar 12, 2020, 8:24 AM Beraldo Leal <notifications@github.com> wrote:

> *@beraldoleal* commented on this pull request.
> ------------------------------
>
> In avocado/core/future/settings.py
> <https://github.com/avocado-framework/avocado/pull/3641#discussion_r391584511>
> :
>
> > +            use the section+key as metavar.
> +
> +        required : bool
> +            If this is a required option or not when on command-line. Default
> +            is False.
> +
> +        action :
> +            The basic type of action to be taken when this argument is
> +            encountered at the command line. For more information visit the
> +            argparser documentation.
> +
> +        .. note:: Most of the arguments here (like parser, positional_arg,
> +                  short_arg, long_arg, choices, nargs, metavar, required and
> +                  action) are only necessary if you would like to add a
> +                  command-line option.
> +        """
>
> Not sure if I got your comment.
>
> IMO, a perfect scenario would be to abstract all the internal details from
> the API user. However, today we are still using choices, nargs, metavar,
> required and action that are required by the argparser and they are there.
>
> On parser argument also I noted the type "argparser parser".
>
> So, I'm not sure what to do here. @clebergnu
> <https://github.com/clebergnu> could you eli5? :)
>
That was just a brain dump, no need to worry about it. Let's move along!

**Comment by beraldoleal** [comment]
@clebergnu all changes are here now. Please see if I'm missing something. Thanks again.

**Comment by beraldoleal** [comment]
@clebergnu closing in favor of v2 (#3644)
